osu.Game/Graphics/Containers/SectionsContainer.cs

```diff
@@ -195,16 +204,49 @@ protected override void UpdateAfterChildren()
                 headerBackgroundContainer.Y = ExpandableHeader?.Y ?? 0;
 
                 float scrollOffset = FixedHeader?.LayoutSize.Y ?? 0;
-                Func<T, float> diff = section => scrollContainer.GetChildPosInContent(section) - currentScroll - scrollOffset;
 
-                if (scrollContainer.IsScrolledToEnd())
+                float visibleHeight(T section)
+                {
+                    float sectionPosition = scrollContainer.GetChildPosInContent(section) - scrollOffset;
+                    float top = Math.Max(sectionPosition, currentScroll);
+                    float bottom = Math.Min(sectionPosition + section.Height, currentScroll + scrollContainer.DisplayableContent);
+                    return bottom - top;
+                }
+
+                var sortedByVisible = new List<T>(Children);
+                sortedByVisible.Sort((a, b) => visibleHeight(b).CompareTo(visibleHeight(a)));
+                T mostVisible = sortedByVisible.FirstOrDefault();
+
+                if (mostVisible == null)
+                    return;
+
+                int mostVisibleIndex = IndexOf(mostVisible);
+
+                bool usePreviousSection()
```

Comment:
You can't just drop inline functions in the middle of a huge block like this. It makes for really bad readability. I'd split this out into its own private method if possible, including documentation on what "usePreviousSection" actually means.